README.md
# Yatzy-app

Sovellus on klassikko noppapeli nimeltä Yatzy. Käyttäjä voi käyttöliittymää käyttäen pelata peliä.

## Dokumentaatio

- [Käyttöohje](./dokumentaatio/kayttoohje.md)
- [Changelog](./dokumentaatio/changelog.md)
- [Tuntikirjanpito](./dokumentaatio/tuntikirjanpito.md)
- [Vaatimusmäärittely](./dokumentaatio/vaatimusmaarittely.md)
- [Arkkitehtuurikuvaus](./dokumentaatio/arkkitehtuuri.md)
- [Testausdokumentti](./dokumentaatio/testaus.md)

## Asennus

1. Riippuvuuksien asentamiseen käytä "yatzy-app" repositorion sisällä komentoa: 
```bash
poetry install
```

2. Ohjelman aloittamiseen käytä "yatzy-app" repositorion sisällä komentoa:
```bash
poetry run invoke start
```

## Komentorivitoiminnot

### Suoritus
```bash
poetry run invoke start
```

### Testaus
```bash
poetry run invoke test
```

### Testikattavuusraportti

```bash
poetry run invoke coverage-report
```

### Pylint

Tiedoston [.pylintrc](./.pylintrc) tarkastukset tehdään "yatzy-app" repositorion sisällä komennolla:

```bash
poetry run invoke lint
```

## Releaset
Releaset löytyvät tästä linkistä:
https://github.com/mleikas/ot-harjoitustyo/releases


dokumentaatio/changelog.md
## VIIKKO 3

- Nopan heiton visuaalinen runko tehty
- Numeroiden arvojen generoiminen tehty
- Testattu, että generoituja nopanheittoja tulee 5

## VIIKKO 4

- Grafiikkaa parannettu
- Lisätty painike nopan pitämiseen
- Valmistauduttu eri noppalogiikan vaihtoon tekemällä niille luokka
- Lisätty muita keskeneräisiä peliin tarvittavia asioita, kuten pisteet ja tekstit kombinaatioille

## VIIKKO 5
- Graafiikkaa parannettu: painikkeet pisteille, väriä
- Sovelluksessa noppia voidaan nyt pitää, heittää 3 kertaa per vuoro, nähdä pisteet, jota saadaan
- Sovelluslogiikkaa vaihdettu luokkiin
- Pisteiden tarkistus
- Pisteiden saanti nopista ja laskeminen
- Pisteiden päivitys ruudulle

## VIIKKO 6
- Lisätty peli ohi viesti
- Pylint korjaukset
- Docstring dokumentaatio

## VIIKKO 7
- Pelin viimeistely: saatavat pisteet ja info
- Funktioiden tekeminen
- Testien tekeminen
- Dokumentaation jatkaminen


yatzy-app/src/tests/dice_test.py
# pylint: disable-all
# pylint: skip-file
import unittest
import class_yatzy, class_dice, class_scores, class_updatescore
class TestDice(unittest.TestCase):
    def setup(self):
        self.remaining_rolls = 1

    def test_dice(self):
        self.assertEqual(Yatzy.roll, None)

yatzy-app/src/tests/yatzy_test.py
# pylint: disable-all
# pylint: skip-file

"""
Lähes mahdotonta testata funktioita, kun ne eivät palauta mitään tai niissä on kiinni myös UI kamaa :/
"""
import unittest
import tkinter as TK
from class_yatzy import Yatzy, count_total
from class_dice import Dice
from class_scores import Scores
from class_updatescore import UpdateScore
import tkinter as TK
from tkinter import messagebox
import random

class TestYatzy(unittest.TestCase):
    def setUp(self):

        scorelist = [1, 2, 3, 4, 5, 6, "Kolme samaa", "Neljä samaa", "Full House",
        "Pieni suora", "Suuri suora", "Yatzy", "Sattuma"]
        self.top_score = 13
        self.bottom_score = 14
        self.values = [5,5,5,5,5]
        self.dice = []

        self.scores = {}
        for i in scorelist:
            self.scores[i] = 1
        self.temp_scores = {}
        for i in scorelist:
            self.temp_scores[i] = 0
        top_score = 55
        bonus = 0
        bottom_score = 50
        self.total = count_total(top_score, bonus, bottom_score)

    def count_total(self, i, j, k):
        pass
    def test_game_over_screen(self):
        """
        Tämä näkyy testin tehdessä testaajalle
        """
        self.assertEqual(Yatzy.game_over(self), None)

    def test_remaining_rolls_after_roll(self):
        self.remaining_rolls = 3
        Yatzy.remaining_rolls_roll(self)
        self.assertEqual(self.remaining_rolls, 2)
    
    def test_total_score(self):
        self.assertEqual(self.total, 105)

    def test_setting_hold_boolean_to_die(self):
        Dice.hold = TK.BooleanVar()
        Dice.number = TK.IntVar()
        Dice.set_hold_and_number(Dice,False, 5)
        self.assertEqual(Dice.hold.get(), False)

    def test_setting_die_number(self):
        Dice.hold = TK.BooleanVar()
        Dice.number = TK.IntVar()
        Dice.set_hold_and_number(Dice,False, 5)
        self.assertEqual(Dice.number.get(), 5)


laskarit/viikko2/unicafe/src/tests/kassapaate_test.py
import unittest
from kassapaate import Kassapaate
from maksukortti import Maksukortti

class TestKassapaate(unittest.TestCase):
    def setUp(self):
        self.kassapaate = Kassapaate()
        self.maksukortti = Maksukortti(1000)
        self.maksukortti2 = Maksukortti(200)
        self.saldo = 1000

    def test_konstruktori_asettaa_saldon_oikein(self):
        self.assertEqual(str(self.kassapaate.kassassa_rahaa), "100000")

    def test_syo_edullisesti_vahentaa_saldoa_oikein(self):
        self.kassapaate.syo_edullisesti_kateisella(self.saldo)
        self.assertEqual(str(self.kassapaate.kassassa_rahaa), "100240")

    def test_syo_maukkaasti_vahentaa_saldoa_oikein(self):
        self.kassapaate.syo_maukkaasti_kateisella(self.saldo)
        self.assertEqual(str(self.kassapaate.kassassa_rahaa), "100400")

    def test_syo_edullisesti_nostaa_myytyjen_maaraa(self):
        self.kassapaate.syo_edullisesti_kateisella(self.saldo)
        self.assertEqual(str(self.kassapaate.edulliset), "1")

    def test_syo_maukkaasti_nostaa_myytyjen_maaraa(self):
        self.kassapaate.syo_maukkaasti_kateisella(self.saldo)
        self.assertEqual(str(self.kassapaate.maukkaat), "1")

    def test_ei_riittavasti_rahaa_ei_muuta_kassan_rahamaaraa(self):
        self.kassapaate.syo_maukkaasti_kateisella(20)
        self.assertEqual(str(self.kassapaate.kassassa_rahaa), "100000")
    
    def test_ei_riittavasti_rahaa_ei_muuta_myytyjen_maaraa(self):
        self.kassapaate.syo_maukkaasti_kateisella(20)
        self.assertEqual(str(self.kassapaate.maukkaat), "0")

    def test_korttiosto_toimii_edullisesti_ja_palautetaan_true(self):
        self.assertEqual(self.kassapaate.syo_edullisesti_kortilla(self.maksukortti), True)

    def test_korttiosto_edullisesti_nostaa_lounaiden_maaraa(self):
        self.kassapaate.syo_edullisesti_kortilla(self.maksukortti)
        self.assertEqual(str(self.kassapaate.edulliset), "1")

    def test_korttiosto_toimii_maukkaasti_ja_palautetaan_true(self):
        self.assertEqual(self.kassapaate.syo_maukkaasti_kortilla(self.maksukortti), True)

    def test_korttiosto_maukkaasti_nostaa_lounaiden_maaraa(self):
        self.kassapaate.syo_maukkaasti_kortilla(self.maksukortti)
        self.assertEqual(str(self.kassapaate.maukkaat), "1")

    def test_saldo_ei_mene_negatiiviseksi_ja_myydyt_pysyvat_samana(self):
        self.kassapaate.syo_maukkaasti_kortilla(self.maksukortti2)  
        self.assertEqual(self.kassapaate.syo_maukkaasti_kortilla(self.maksukortti2), False)           
        self.assertEqual(str(self.kassapaate.maukkaat), "0")
    
    def test_saldo_kortissa_ja_kassassa_muuttuvat(self):
        self.kassapaate.lataa_rahaa_kortille(self.maksukortti, 200)
        self.assertEqual(self.maksukortti.saldo, 1200)
        self.assertEqual(str(self.kassapaate.kassassa_rahaa), "100200")

yatzy-app/src/class_dice.py
"""
Noppien luokka
"""

import tkinter as TK

class Dice(TK.Checkbutton):
    """Luokka, jonka avulla määritetään nopat.

    Tähän kuuluu nopan silmäluvun muuttuja ja pidetäänkö noppaa vai ei

    """

    def __init__(self, master):
        """Luokan konstruktori, joka luo uuden nopan

        """
        self.master = master
        self.hold = TK.BooleanVar()
        self.number = TK.IntVar()
        self.set_hold_and_number(True, 0)
        self.off_color = TK.PhotoImage(width=300,height=50)
        self.on_color = TK.PhotoImage(width=300,height=50)
        self.off_color.put(("green",), to=(0,0,80,50))
        self.on_color.put(("red",), to=(0,0,80,50))
        TK.Checkbutton.__init__(self, master=master, font=("Helvetica",30),
        image=self.off_color, selectimage=self.on_color, var=self.hold)

    def set_hold_and_number(self, boolean, num):
        """
        Laittaa arvot nopan pitämiselle ja silmäluvulle
        """
        self.hold.set(boolean)
        self.number.set(num)


yatzy-app/src/class_scores.py
"""
Luokka kombinaatiopainikkeille ja niiden antamille pisteille
"""
import tkinter as TK

top_list = [1,2,3,4,5,6]
teksti = ["-","Ykköset","Kakkoset","Kolmoset","Neloset","Viitoset","Kuutoset"]
class Scores(TK.Frame):
    """Luokka, jonka avulla hoidetaan painikkeet pisteiden saantia varten

    Attributes:
        score: Pisteiden määrä
        name: Painikkeen nimi
    """
    def __init__(self, master, name, score=None):
        """Luokan konstruktori, joka luo painikkeen pisteille ja pistemuuttujan,
        joka pitää kirjaa painikkeen pisteistä

        Args:
            score: Pisteiden määrä
            name: Painikkeen nimi
        """
        self.master = master
        TK.Frame.__init__(self, master)
        text = name
        if name in top_list:
            text = teksti[name]
        self.name = name
        self.score = score
        self.var_score = TK.IntVar()
        self.button = TK.Button(self, text=text, command=self.press, font=("Helvetica",15))
        self.button.pack(side="left")
        self.label = TK.Label(self, textvar=self.var_score)
        self.label.pack(side="left")

    def press(self):
        """Painikketta painaessa tallentaa saadut pisteet ja muuttaa painikkeen väriä

        Lisäksi, jos kaikki painikkeet ovat painettu, peli loppuu

        """
        if self.score is None and 0 not in [i.number.get() for i in self.master.dice]:
            self.master.scores[self.name] = self.master.temp_scores[self.name]
            self.score = self.master.temp_scores[self.name]
            self.var_score.set(self.score)
            for i in self.master.dice:
                i.number.set(0)
                i.hold.set(True)
            if self.name in top_list:
                self.master.top_score += self.score
            else:
                self.master.bottom_score += self.score
            self.button.config(bg="green")
            self.label.config(bg="green")
            self.master.update_score()
            self.master.remaining_rolls = 3

            if None not in self.master.scores.values():
                self.master.game_over()


yatzy-app/src/class_updatescore.py
"""
Päivittää käyttöliittymään pisteiden määrän
"""

import tkinter as TK


class UpdateScore(TK.Frame):
    """
    Luokka, joka näyttää ruudulla yläpisteet, alapisteet ja kokonaispisteet
    """
    def __init__(self, master):
        self.master = master
        TK.Frame.__init__(self, master)
        self.var_top_score = TK.IntVar()
        self.var_top_score.set(self.master.top_score)
        self.var_bottom_score = TK.IntVar()
        self.var_bottom_score.set(self.master.bottom_score)
        self.var_bonus = TK.IntVar()
        self.var_bonus.set(0)
        self.var_sum_score = TK.IntVar()
        self.var_sum_score.set(self.master.top_score+self.master.bottom_score+self.var_bonus.get())
        TK.Label(self, text="Vasemman puolen pisteet").pack(side="top")
        TK.Label(self, textvar=self.var_top_score).pack(side="top")
        TK.Label(self, text="Oikean puolen pisteet").pack(side="top")
        TK.Label(self, textvar=self.var_bottom_score).pack(side="top")
        TK.Label(self, text="Kokonaispisteet").pack(side="top")
        TK.Label(self, textvar=self.var_sum_score).pack(side="top")


yatzy-app/src/class_yatzy.py
"""
Pelin luokka
"""

import tkinter as TK
from tkinter import messagebox
import random
from class_dice import Dice
from class_scores import Scores
from class_updatescore import UpdateScore


dice = ["\u2610","\u2680","\u2681","\u2682","\u2683","\u2684","\u2685"]
scorelist = [1, 2, 3, 4, 5, 6, "Kolme samaa", "Neljä samaa", "Full House",
"Pieni suora", "Suuri suora", "Yatzy", "Sattuma"]

class Yatzy(TK.Frame):
    """Luokka, joka ylläpitää peliä

    """
    def __init__(self, master):
        TK.Frame.__init__(self, master, bg="#9F561F", borderwidth=50)
        self.master = master
        self.scores = {}
        for i in scorelist:
            self.scores[i] = None
        self.dice = []
        for i in range(5):
            self.dice.append(Dice(self))
        self.top_score = 0
        self.bottom_score = 0
        self.temp_scores = {}
        for i in scorelist:
            self.temp_scores[i] = 0
        self.remaining_rolls = 3
        self.receivable_points = TK.StringVar(self, "Saatavat pisteet: -")
        for i, die in enumerate(self.dice):
            die.grid(row=i, column=0, padx=1, pady=1,sticky="E")
            TK.Label(self, font=("Helvetica",30), text="Pidä Noppa").grid(
                row=i, column=0, padx=5, pady=1,sticky="E")
        self.update_dice_art()
        TK.Button(self,text="Heitetään noppaa!", font=("Helvetica",30),
            foreground="green",command=self.roll).grid(row=i+1, column=0, rowspan=2)
        row = 0
        col = 1
        self.score_table = []
        for i, j in self.scores.items():
            frame = Scores(self, i, j)
            self.score_table.append(frame)
            frame.grid(row=row, column=col+1,sticky="S", padx=20)
            row += 2
            if row == 6:
                row = 0
                col += 1
        self.updatescore = UpdateScore(self)
        self.updatescore.grid(row=row, column=col+1,sticky="S",pady=10, padx=40)
        TK.Label(self, text="Pidä Noppa: vihreä pitää nopan, punainen heittää",
            font=("Helvetica",11)).grid(row=row+5, column=0, padx=1,pady=1)

    def check_scores(self):
        """Tarkistetaan, paljonko pisteitä saadaan kierroksen nopista

        """
        for i in scorelist:
            self.temp_scores[i] = 0
        values = []
        for i in self.dice:
            values.append(i.number.get())
        values.sort()
        for i in range(1,7):
            self.temp_scores[i] = sum([x for x in values if x == i])
            if values.count(i) >= 3:
                self.temp_scores["Kolme samaa"] = sum(values)
                if len(set(values)) == 2:
                    self.temp_scores["Full House"] = 25
            if values.count(i) >= 4:
                self.temp_scores["Neljä samaa"] = sum(values)
            if values.count(i) == 5:
                self.temp_scores["Yatzy"] = 50
        str_check = "".join(sorted(list(set([str(a) for a in values]))))
        for j in ["1234","2345","3456"]:
            if j in str_check:
                self.temp_scores["Pieni suora"] = 30
                break
        if str_check in ["12345","23456"]:
            self.temp_scores["Suuri suora"] = 40
        self.temp_scores["Sattuma"] = sum(values)

    def roll(self):
        """Antaa nopille silmäluvut ja päivittää tiedot

        Tarkistaa kuinka monta heittoa jäljellä

        """
        if self.remaining_rolls > 0:
            for i in self.dice:
                if i.hold.get() or not i.number.get():
                    i.number.set(random.randrange(1,7))
            self.check_scores()
            self.available_points()
            self.update_dice_art()
            self.remaining_rolls_roll()
            TK.Label(self, text="Heittoja jäljellä: " + str(self.remaining_rolls),
            font=("Helvetica",15)).grid(row=5, column=1, sticky="S", padx=20, pady=40)

    def remaining_rolls_roll(self):
        """
        Vähentää jäljellä olevia heittoja
        """
        self.remaining_rolls -= 1

    def available_points(self):
        """
        Vaihtaa saatavien pisteiden numerot
        """
        row = 0
        col = 1
        for points in self.temp_scores.values():
            self.receivable_points = TK.Label(self, text="Saatavat pisteet: " + str(points),
            font=("Helvetica",13), width=16)
            self.receivable_points.grid(row=row+1, column=col+1,sticky="S", padx=10,pady=50)
            row+=2
            if row == 6:
                row = 0
                col += 1

    def game_over(self):
        """Tuottaa peli loppui viestin pelaajalle

        """
        if None not in self.scores.values():
            top_score = sum([self.scores[i] for i in scorelist[:6]])
            if top_score >= 63:
                bonus = 50
            else:
                bonus = 0
            bottom_score = sum([self.scores[i] for i in scorelist[6:]])
            total = count_total(top_score, bonus, bottom_score)

            message1 = f"Vasemman puolen pisteet: {top_score}\n Bonus: \
            {bonus}\n Oikean puolen pisteet: {bottom_score}\n Kokonaispisteet: {total}"
            TK.messagebox.showinfo(title="Peli loppui!", message=message1)

    def update_score(self):
        """
        Asettaa pisteet pisteiden päivitystä varten
        """
        self.updatescore.var_top_score.set(self.top_score)
        self.updatescore.var_bottom_score.set(self.bottom_score)
        self.updatescore.var_sum_score.set(self.top_score+self.bottom_score)

    def update_dice_art(self):
        """Päivittää noppien grafiikat

        """
        for i, die in enumerate(self.dice):
            self.dice_art = TK.Label(self, text=dice[die.number.get()], width=2,
            textvar=dice[die.number.get()],fg="black",bg="#9F561F",font=("Helvetica",75))
            self.dice_art.grid(row=i, column=1)

    def __call__(self):
        """Suorittaa tkinterille annetut käskyt

        """
        self.pack()
        self.mainloop()

def count_total(top, bonus, bot):
    """
    Laskee pisteet yhteen
    """
    return top + bonus + bot
